walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,94 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
 
 public final class FindMeetingQuery {
-  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
-  }
+    /**
+     * Given a set of events and a list of attendees, find and return the set of
+     * valid meeting times.
+     */
+    public Collection<TimeRange> findMeetingTimes (
+        Collection<Event> events,    Collection<String> attendees,
+        long duration) {
+
+        /* Sort events with latest end times coming first. */
+        List<Event> eventsByEndTime = new ArrayList<Event>(events);
+        List<Event> eventsByStartTime = new ArrayList<Event>(events);
+        Collections.sort(eventsByEndTime, Event.ORDER_BY_END_DESCENDING);
+        Collections.sort(eventsByStartTime, Event.ORDER_BY_START_ASCENDING);
+
+        Collection<TimeRange> meetingTimes = new ArrayList<TimeRange>();
+        int prevEndTime = TimeRange.END_OF_DAY + 1;
+        
+        for (Event event : eventsByEndTime) {
+            /* Find attendees required to be at meeting also attending event. */
+            Collection<String> conflictAttendees = new HashSet<>(attendees);
+            Collection<String> eventAttendees = event.getAttendees();
+            conflictAttendees.retainAll(event.getAttendees());
+
+            /**
+             * For events that have conflicting attendees, add the time range 
+             * spanning from the last recorded end time to the current event's end 
+             * time to our output set if the duration fits within it. Move last end 
+             * time to current event's start time for the next event. 
+             */
+            if (conflictAttendees.size() > 0) {
+                TimeRange freeTime = TimeRange.fromStartEnd(event.getWhen().end(), prevEndTime, false);
+                if (duration <= freeTime.duration()) {
+                    meetingTimes.add(freeTime);
+                } 
+                prevEndTime = (event.getWhen().start() < prevEndTime) ? event.getWhen().start() : prevEndTime;
+            }
+        }
+
+        /* Check the time range from the start of the day to the first event's start time. */
+        if (prevEndTime > 0) {
+            TimeRange firstFreeTime = TimeRange.fromStartEnd(0, prevEndTime, false);
+            meetingTimes.add(firstFreeTime);
+        }
+
+        List<TimeRange> meetingTimesList = new ArrayList<TimeRange>(meetingTimes);
+        Collections.reverse(meetingTimesList);
+        return meetingTimesList;
+    }
+
+    /**
+     * The query function returns a set of TimeRanges where a potential meeting,
```

Comment:
It's unclear if the `Collection<TimeRange>` will contain overlapping ranges or not.